indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/Asn1Elements.kt

```diff
@@ -78,10 +81,27 @@ sealed class Asn1Element(
 
     val tag by lazy { tlv.tag }
 
-    val derEncoded: ByteArray by lazy {
-        children?.fold(byteArrayOf()) { acc, extendedTlv -> acc + extendedTlv.derEncoded }
-            ?.let { byteArrayOf(*tlv.tag.encodedTag, *it.size.encodeLength(), *it) }
-            ?: byteArrayOf(*tlv.tag.encodedTag, *encodedLength, *tlv.content)
+
+    private val derEncodedLazy = lazy { Buffer().also { it.writeAsn1Element(this) }.readByteArray() }
+    val derEncoded: ByteArray by derEncodedLazy
+
+    private fun Sink.writeAsn1Element(element: Asn1Element) {
+        if (element.derEncodedLazy.isInitialized()) {
+            write(element.derEncoded)
+            return
+        }
+
+        element.children?.let { childElems ->
+            write(element.tlv.tag.encodedTag);
+            write(element.encodedLength);
+            childElems.forEach { child -> writeAsn1Element(child) }
+        }
+
+            ?: also { //primitive
+                write(element.tlv.tag.encodedTag)
+                write(element.encodedLength)
+                write(element.tlv.content)
+            }
```

Comment:
for the record i still think that we should revisit the entire `tlv` concept, since it seems that for any given element either "elm.children" or "elm.tlv.content" is used, which sounds like the two classes are so intrinsically linked they should maybe be a single class. At the very least the "has children" vs "has content" distinction should be a sealed-class diff.

either way that's a change for another PR.